doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,299 @@
+# Reclocking implementation
+
+## Summary
+
+This document proposes a design for implementing reclocking and exposing the
+reclocking mappings to end users.
+
+## Goals
+
+* Describe the physical representation of the reclocking data.
+* Describe the policy for reclocking each type of source.
+* Describe how users interact with the reclocking data.
+
+## Non-Goals
+
+* Motivating reclocking. See the product spec in [#11218].
+* Describing reclocking formally. See the
+  [original reclocking design doc](20210714_reclocking.md).
+
+## Description
+
+### Reclockable source types
+
+We define a **reclockable** source type as one where the upstream system
+provides a useful [gauge of progress](20210714_reclocking.md#description).
+The following source types are reclockable:
+
+* File
+* Kafka
+* Kinesis
+* PostgreSQL
+* S3
+
+The following sources types are nonreclockable:
+
+* PubNub
+
+### Storage layer implementation
+
+A reclockable source will consist of two definite collections in the storage
+layer: the [`remap`](20210714_reclocking.md#remap) collection, which contains
+the reclocking metadata, and the reclocked data collection.
+
+Nonreclockable sources will consist of only a single definite collection
+containing the source's data. For consistency with reclockable sources, we will
+refer to this collection as the "reclocked" collection, even though there is no
+upstream clock.
+
+The type of the `remap` collection depends on the source type:
+
+| Source type | `remap` type                                     |
+|-------------|--------------------------------------------------|
+| File        | `(ignored: (), byte_offset_increment: i64)`      |
+| Kafka       | `(partition: i32, offset_increment: i64)`        |
+| Kinesis     | `(shard: i64, sequence_increment: numeric)`      |
+| PostgreSQL  | `(ignored: (), lsn_increment: i64)`              |
+| S3          | `(object_key: text, byte_offset_increment: i64)` |
+
+The `remap` collection exists in the Materialize time domain, where a `u64`
```

Comment:
((You added the "mechanics" section after I commented about "maybe having insights" on Slack. 😅))

This comment is mostly relevant for an alternative implementation strategy. In that alternative strategy, minting of bindings happens "inline" with reading the data, because that is something most sources support and doesn't rely on there being an API call for getting the latest offsets. Another potential advantage of this is that source readers don't have to wait for the `remap` collection to advance before data can be timestamped, because they actively participate.

When using `ToTime` as the `ts` of the binding updates (as the timestamp of that collection) it means that all bindings that ever will be minted for a given `ToTime` `t` need to be written as one batch, by one writer/worker. This follows from how `Append` works in `formalism.md` and in turn how the new persist `append` API works. That's fine if there is one central entity that mints bindings for all partitions/shards and if we only mint new bindings when `current_time` "ticks" (rolls over to the new `ToTime`, say when you have a 1-second timestamp interval that happens once every second).

In a model source pipeline that we implemented we use a sequence number as the timestamp that drives the bindings collection and multiple (potentially concurrent source readers) mint new bindings whenever they see source offsets that are not yet covered. Internally, this uses a CaS-like operation to ensure that concurrent readers agree on minted bindings. Decoupling the timestamp that drives the collection from `ToTime` allows writing bindings for the same `ToTime` at different points in time.

Reply:
This alternative checks out for me. I have to confess I am wholesale laundering @frankmcsherry's opinion in the "mechanics" section. @frankmcsherry—what's your take here?

Reply:
I think I probably didn't understand @aljoscha completely. The last paragraph definitely went over my head. But here are some takes on the first paragraphs:

> Another potential advantage of this is that source readers don't have to wait for the remap collection to advance before data can be timestamped, because they actively participate.

I think they do have to wait for it to advance in either case, as the source should not publish any data until the `remap` collection is durable. But now we are in a state where there are potentially multiple source instances reading and creating bindings who now need to reconcile their takes (e.g. if we are failing over from one source reader to another). This consensus logic is now in the middle of the implementation of the reader, and likely harder to adjust or adapt.

> minting of bindings happens "inline" with reading the data, because that is something most sources support and doesn't rely on there being an API call for getting the latest offsets.

At the risk of proposing something too exotic, you can read a stream once and produce two outputs: the bindings, and the data that are associated with those bindings. We don't have to throw away the data when we read the stream to first produce the bindings. But we also don't have to fuse the reclocking and data production operators. Keeping them separate seems like it can allow less fraught implementations (though perhaps I don't understand your model: can you explain what it would look like to ingest a 64 part Kafka topic on 64+ workers with high availability?).